packages/context/src/__tests__/unit/handler-chain.unit.ts

```diff
@@ -0,0 +1,191 @@
+      'res-metadata-handler1',
+    ]);
+  });
+
+  it('catches error from second handler', async () => {
+    givenInvocationHandlerChain({
+      handler1: {throwError: false},
+      handler2: {throwError: true},
+    });
+    const resultPromise = handlerChain.invokeHandlers();
+    await expect(resultPromise).to.be.rejectedWith('error in handler2');
+    expect(events).to.eql(['before-handler1', 'before-handler2']);
+  });
+
+  it('catches error from first handler', async () => {
+    givenInvocationHandlerChain({
+      handler1: {throwError: true},
+      handler2: {throwError: false},
+    });
+    const resultPromise = handlerChain.invokeHandlers();
+    await expect(resultPromise).to.be.rejectedWith('error in handler1');
+    expect(events).to.eql([
+      'before-handler1',
+      'before-handler2',
+      'after-handler2',
+    ]);
+  });
+
+  it('allows discovery of handlers in context', async () => {
+    const handler1 = givenHandler('handler1');
+    const handler2 = givenHandler('handler2');
+    ctx
+      .bind('handler2')
+      .to(handler2)
+      .tag('handler');
+    ctx
+      .bind('handler1')
+      .to(handler1)
+      .tag('handler');
+    handlerChain = new InvocationHandlerChain(ctx, filterByTag('handler'));
+    await handlerChain.invokeHandlers();
+    expect(events).to.eql([
+      'before-handler2',
+      'before-handler1',
+      'after-handler1',
+      'after-handler2',
+    ]);
+  });
+
+  it('allows discovery and sorting of handlers in context', async () => {
+    const handler1 = givenHandler('handler1');
+    const handler2 = givenHandler('handler2');
+    ctx
+      .bind('handler2')
+      .to(handler2)
+      .tag('handler')
+      .tag({phase: 'p2'});
+    ctx
+      .bind('handler1')
+      .to(handler1)
+      .tag('handler')
+      .tag({phase: 'p1'});
+    handlerChain = new InvocationHandlerChain(
+      ctx,
+      filterByTag('handler'),
+      compareBindingsByTag('phase', ['p1', 'p2']),
+    );
+    await handlerChain.invokeHandlers();
+    expect(events).to.eql([
+      'before-handler1',
+      'before-handler2',
+      'after-handler2',
+      'after-handler1',
+    ]);
+  });
+
+  function givenContext() {
+    events = [];
+    ctx = new Context();
+  }
+
+  type HandlerOptions = {
+    // The handler won't call `next()`
+    skipNext?: boolean;
+    // Set a value for the handler to return
+    returnValue?: InvocationResult;
+    // The handler will bind metadata to context
+    bindMetadata?: boolean;
+    // The handler will throw an error
+    throwError?: boolean;
+  };
+
+  function givenInvocationHandlerChain(
+    options: {
+      handler1: HandlerOptions;
+      handler2: HandlerOptions;
+    } = {handler1: {}, handler2: {}},
+  ) {
+    const handler1 = givenHandler('handler1', options.handler1);
+    const handler2 = givenHandler('handler2', options.handler2);
+    ctx.bind('handler2').to(handler2);
+    const handlers = [handler1, 'handler2'];
```

Comment:
IIUC, this is verifying that handlers can be specified both by a value and by binding key.

I find the current code very problematic:
 - The assertion/expectation is implicit, it's not obvious from the test cases.
 - We are verifying this behavior in every test, which is rather redundant.
 - It makes it difficult to see what the tests are about.